core/src/com/unciv/ui/tilegroups/TileGroupIcons.kt

```diff
@@ -49,46 +42,54 @@ class TileGroupIcons(val tileGroup: TileGroup){
         populationIcon = null
     }
 
+    private fun updateUnitIcon(tileIsViewable: Boolean, showMilitaryUnit: Boolean, viewingCiv: CivilizationInfo?) {
+        civilianUnitIcon?.remove()
+        militaryUnitIcon?.remove()
+        aircraftUnitIcon?.remove()
+        if (tileGroup.tileInfo.civilianUnit != null && tileIsViewable) {
+            civilianUnitIcon = unitIcon(tileGroup.tileInfo.civilianUnit!!, viewingCiv)
+            tileGroup.unitLayerGroup.addActor(civilianUnitIcon)
+        }
+        if (tileGroup.tileInfo.militaryUnit != null && tileIsViewable && showMilitaryUnit) {
+            militaryUnitIcon = unitIcon(tileGroup.tileInfo.militaryUnit!!, viewingCiv)
+            tileGroup.unitLayerGroup.addActor(militaryUnitIcon)
+        }
+        if (tileGroup.tileInfo.airUnits.isNotEmpty()) {
+            aircraftUnitIcon = aircraftIcon(viewingCiv)
+            tileGroup.unitLayerGroup.addActor(aircraftUnitIcon)
+        }
+    }
 
-    fun newUnitIcon(unit: MapUnit?, oldUnitGroup: UnitGroup?, isViewable: Boolean, yFromCenter: Float, viewingCiv: CivilizationInfo?): UnitGroup? {
-        var newImage: UnitGroup? = null
-        // The unit can change within one update - for instance, when attacking, the attacker replaces the defender!
-        oldUnitGroup?.remove()
-
-        if (unit != null && isViewable) { // Tile is visible
-            newImage = UnitGroup(unit, 25f)
-            if (UncivGame.Current.settings.continuousRendering && oldUnitGroup?.blackSpinningCircle != null) {
```

Comment:
I don't rewrite Line 60-63 because I think it is meaningless.
Of course, it's up to you.